File: Unit-1-Divide-and-Conquer/01-Binary-Search-and-Power/binary_search_power.py
from typing import List


def search(nums: List[int], target: int) -> int:
    left = 0
    right = len(nums) - 1

    while left <= right:
        mid = (left + right) // 2

        if nums[mid] == target:
            return mid
        elif nums[mid] < target:
            left = mid + 1
        else:
            right = mid - 1

    return -1


def myPow(x: float, n: int) -> float:
    if n == 0:
        return 1.0

    if n < 0:
        x = 1 / x
        n = -n

    result = 1.0

    while n > 0:
        if n % 2 == 1:
            result *= x

        x *= x
        n //= 2

    return result

def main():
    print("Binary Search")
    nums = list(map(int, input("Enter sorted array elements: ").split()))
    target = int(input("Enter target: "))

    index = search(nums, target)

    if index != -1:
        print("Target found at index:", index)
    else:
        print("Target not found")

    print("\nFast Power")
    x = float(input("Enter base: "))
    n = int(input("Enter exponent: "))

    result = myPow(x, n)

    print("Result:", result)


if __name__ == "__main__":
    main()

File: Unit-1-Divide-and-Conquer/01-Binary-Search-and-Power/binary_search_power.cpp
#include <vector>
#include <iostream>
using namespace std;

int search(vector<int>& nums, int target) {
    int left = 0;
    int right = nums.size() - 1;

    while (left <= right) {
        int mid = left + (right - left) / 2;

        if (nums[mid] == target) {
            return mid;
        }
        else if (nums[mid] < target) {
            left = mid + 1;
        }
        else {
            right = mid - 1;
        }
    }

    return -1;
}

double myPow(double x, int n) {
    if (n == 0) {
        return 1.0;
    }

    long long exponent = n;

    if (exponent < 0) {
        x = 1 / x;
        exponent = -exponent;
    }

    double result = 1.0;

    while (exponent > 0) {
        if (exponent % 2 == 1) {
            result *= x;
        }

        x *= x;
        exponent /= 2;
    }

    return result;
}


int main() {
    vector<int> nums = {1, 3, 5, 7, 9};

    cout << "Index: " << search(nums, 7) << endl;
    cout << "Power: " << myPow(2, 10) << endl;

    return 0;
}